src/driver/include/driver/workload_parsers.hpp

```diff
@@ -15,16 +15,72 @@
 #ifndef HEADER_0369D27D_9A68_4981_B344_4BAB3EE09A80_INCLUDED
 #define HEADER_0369D27D_9A68_4981_B344_4BAB3EE09A80_INCLUDED
 
+#include <optional>
 #include <boost/filesystem.hpp>
 #include <yaml-cpp/yaml.h>
 
 #include <driver/v1/DefaultDriver.hpp>
 
 namespace genny::driver::v1 {
 
-using YamlParameters = std::map<std::string, YAML::Node>;
 namespace fs = boost::filesystem;
 
+/**
+ * Values stored in a Context are tagged with a type to ensure
+ * they aren't used incorrectly.
+ *
+ * If updating, please add to typeNames in workload_parsers.cpp
+ * Enums don't have good string-conversion or introspection :(
+ */
+enum class Type {
+    kParameter,
+    kActorTemplate,
+    kActorInstance,
+};
+
+/**
+ * Manages scoped context for stored values.
+ *
+ * Only create using a ContextGuard.
+ */
+class Context {
+public:
+    using ContextMap = std::map<std::string, std::pair<YAML::Node, Type>>;
```

Comment:
I started doing this, but ended up having to do `std::make_pair` every single time `insert` was called, or do `returnVal->first` every single time `get` was called (the type is redundant for `get`, since you already assert the type by passing it in, so it'll literally always be `returnVal-first`.) That seemed a lot less ergonomic to me.

Instead I used `ContextValue` internally to `Context`, which hopefully at least makes it clearer. Please push back if you think using it in `get` or `insert`'s signature is also worth doing!